Focused Launch: highlight selected plan in focused launch - detailed plans

alshakero opened the pull request:
### Changes proposed in this Pull Request

* In short, fixes #47814.

### Testing instructions

**Setup:**
1. Checkout this branch
2. Run `yarn start` in Claypso
3. Run `yarn dev --sync` in `apps/editing-toolkit`
4. Sandbox an unlaunched site on _a Free plan_ created via `/start`

**Test changes:**
1. Go to `/page/UNLAUNCHED_SITE_ID_CREATED_WITH_START.wordpress.com/home?flags=create/focused-launch-flow`
2. Click on "Launch" to open the Focused Launch modal
3. Select a plan from the Summary view (e.g. Premium plan)
4. Click on "View all plans"
5. Check that the selected plans is correctly highlighted in the plans grid as per #47814

Fixes #47814

alshakero committed:
Highlight selected plan in focused launch

razvanpapadopol commented on packages/plans-grid/src/plans-table/style.scss:
```diff
@@ -241,10 +241,30 @@
+		// popular selected items don't need a blue border
+		&:not( .is-popular ) {
+			border: 1px solid var( --studio-blue-40 );
+			color: var( --studio-blue-40 );
```
I've tested so far only in Focused Launch and I see different colors for popular plan button (`--wp-admin-theme-color` seems to be the default bg color for primary button) and the selected plan button. 

So this AC is not ✅ yet:

> The 2 changes above should also override any other button styles (including the primary button used in the "popular" plan)

tjcafferkey commented:
I don't know if this is the same thing that @razvanpapadopol has mentioned but the following A/C is not met:

> Use `#FAFCFE` as the background color

**Development**
![Screenshot 2021-01-05 at 08 43 20](https://user-images.githubusercontent.com/8639742/103625076-56506300-4f32-11eb-8daa-8c1588d11565.png)

**Design**
![Screenshot 2021-01-05 at 08 43 32](https://user-images.githubusercontent.com/8639742/103625087-5a7c8080-4f32-11eb-9889-de25de54e8ff.png)

razvanpapadopol commented:
> I don't know if this is the same thing that @razvanpapadopol has mentioned but the following A/C is not met:
> 
> > Use `#FAFCFE` as the background color

Yes, this is related to what I've mentioned. It's not only the background but also the primary color that doesn't match if `--wp-admin-theme-color` is different than `--studio-blue-40`.

ciampo commented:
> #### Testing instructions
> * Create or reuse a free unlaunched site.
>   
>   * Let's say its called **http://test123_wordpress_dotcom**.
> * Go to **http://calypso.localhost:3000/domains/add/test123_wordpress_dotcom**.

Maybe I'm doing something wrong, but when I follow these instructions, I see a page where I can choose a domain, and not a page where I can pick a plan

![Screenshot 2021-01-05 at 15 16 35](https://user-images.githubusercontent.com/1083581/103656540-0214a580-4f69-11eb-967d-5af3ab2bf7ba.png)

razvanpapadopol commented:
> Maybe I'm doing something wrong, but when I follow these instructions, I see a page where I can chose a domain, and not a page where I can pick a plan

I believe the correct URL should be `https://wordpress.com/plans/{SITE_SLUG}`. Updating description now.

ciampo commented:
> I believe the correct URL should be https://wordpress.com/plans/{SITE_SLUG}. Updating description now.

Thank you!

>* Create or reuse a free unlaunched site and sandbox the site.
>* Go to `wordpress.com/plans/{SITE_SLUG}`.
>* Select a non-popular plan and purchase it.

Now that i think about it.. I'm not sure it's necessary to pre-purchase a plan in order to test for these changes? I think the changes can be tested just by opening Focused Launch flow, visiting plan details view, and selecting a plan.

If a paid plan is already associated to the site when opening Focused Launch, the plan option should be greyed out, right?

razvanpapadopol commented:
> Now that i think about it.. I'm not sure it's necessary to pre-purchase a plan in order to test for these changes? I think the changes can be tested just by opening Focused Launch flow, visiting plan details view, and selecting a plan.
> 
> If a paid plan is already associated to the site when opening Focused Launch, the plan option should be greyed out, right?

That's correct.

alshakero committed:
Override wp-admin primary color and add BG

alshakero committed:
Fix the blue variant for focused launch

alshakero committed:
Add new line

alshakero committed:
Fix misreplaced border color

alshakero commented:
I think this is ready for another round:

- [x] Read " ✔️ Current Selection"
- [x] Use `#FAFCFE` as the background color (we should try to use a pre-existing color variable, if there's one that matches)
- [x] Use the primary blue color for borders and text
- [x] The 2 changes above should also override any other button styles (including the primary button used in the "popular" plan)

ciampo commented on packages/launch/src/_variables.scss:
```diff
@@ -6,5 +6,4 @@ $launch-padding-h-wide: 80px;
 
 // @TODO: refactor to common variable in onboarding
 // see https://github.com/Automattic/wp-calypso/issues/47181
-$focused-launch-accent-blue: #117ac9;
```
Can we also move the TODO comment to the onboarding package? I'm afraid here it will get lost / out of context, without being next to this variable

ciampo commented on packages/onboarding/styles/variables.scss:
```diff
@@ -18,3 +18,6 @@ $onboarding-heading-margin-medium: 80px 0 ( 80px * 0.8 );
 $onboarding-body-margin-medium: 0 0 80px 0;
+
+// the blue accent this is used as the primary color in focused-launch
+$focused-launch-accent-blue: #117ac9;
```
This is definitely a better spot for this variable, at least for now (I'm reading the conversation about moving it to the color-studio repo eventually).

I think that we could still give it a better name, since it looks like this variable is meant to be used across different flows (step by step launch, gutenboarding, focused launch).. maybe `$onboarding-accent-blue`? What do you think?

StefanNieuwenhuis replied:
I've renamed the variable 👍

ciampo commented on packages/plans-grid/src/plans-table/style.scss:
```diff
@@ -234,17 +235,43 @@
 	}
 }
 
+// it's OK to use $focused-launch-accent-blue variable here because the full-width mode is only used there
```
This approach seems a bit brittle and it may confuse other developers, since it makes the button look different only based on the `.full-width` class.

It's not an issue created by this PR (and therefore not a blocker), but more of a sign that the plans grid package needs to be refactored and tidied up before it becomes too messy to maintain.

StefanNieuwenhuis replied:
I've refactored the styling for the button states.

ciampo reviewed (changes requested):
The following A/C doesn't seem to be met:

> The 2 changes above should also override any other button styles (**including the primary button used in the "popular" plan**)

To replicate, visit the Plan Details view (in the Focused Launch modal) and compare the styles of the selected plan's button, when the Personal plan and the Premium plan are selected.

In particular, when the Premium plan (the popular one) is selected, the button text updates correctly, but the button's styles are the same as when the plan is not selected.

| Personal plan selected | Premium plan selected |
|---|---|
| <img width="1495" alt="Screenshot 2021-01-08 at 14 42 09" src="https://user-images.githubusercontent.com/1083581/104024158-f6242000-51c2-11eb-82f9-28b1e6399e2a.png"> | <img width="1500" alt="Screenshot 2021-01-08 at 14 42 22" src="https://user-images.githubusercontent.com/1083581/104024153-f4f2f300-51c2-11eb-8a49-5cb3fb58f157.png"> |


I also left a couple of minor comments.

StefanNieuwenhuis committed:
Rename $focused-launch-accent-blue to $onboarding-accent-blue

StefanNieuwenhuis committed:
Move todo-item to onboarding package

We want to move the @TODO to prevent it from getting lost or being taken out of context.

StefanNieuwenhuis committed:
Update button state styling

StefanNieuwenhuis committed:
Remove old button state styles

ciampo reviewed (changes requested):
>Same as [previous review](https://github.com/Automattic/wp-calypso/pull/48622#pullrequestreview-564273094) — when the popular/recommended plan (i.e. premium) is selected, the button's styles are the same as when the plan is not selected — instead they should be overridden by the "selected" styles.

I've updated the  popular/recommended plan button's selected state to match AC.

StefanNieuwenhuis committed:
Update primary button selected style

ciampo commented:
Since monthly plans have been released, we have to check if a plan is currently selected via its product ID — I'll go ahead and make those changes myself

ciampo committed:
Use productIds (and billingPeriod) to check if a plan is selected

ciampo commented on packages/plans-grid/src/plans-accordion/index.tsx:
```diff
@@ -59,7 +59,13 @@ const PlansAccordion: React.FunctionComponent< Props > = ( {
 
 	// Primary plan
-	const popularPlan = useSelect( ( select ) => select( PLANS_STORE ).getDefaultPaidPlan( locale ) );
+	const { popularPlan, getPlanProduct } = useSelect( ( select ) => {
+		const plansStore = select( PLANS_STORE );
+		return {
+			popularPlan: plansStore.getDefaultPaidPlan( locale ),
+			getPlanProduct: plansStore.getPlanProduct,
+		};
+	} );
```
Getting the `getPlanProduct` function from the store

ciampo commented on packages/plans-grid/src/plans-accordion/index.tsx:
```diff
@@ -127,7 +133,8 @@ const PlansAccordion: React.FunctionComponent< Props > = ( {
 								isPrimary
 								isSelected={
 									!! selectedPlanProductId &&
-									primaryPlan.productIds.indexOf( selectedPlanProductId ) > -1
+									selectedPlanProductId ===
+										getPlanProduct( primaryPlan.periodAgnosticSlug, billingPeriod )?.productId
```
comparing with the correct `productId` (and making sure they're not both `undefined`)

ciampo commented on packages/plans-grid/src/plans-accordion/index.tsx:
```diff
@@ -164,7 +171,11 @@ const PlansAccordion: React.FunctionComponent< Props > = ( {
 									openPlans.indexOf( plan.periodAgnosticSlug ) > -1 &&
 									! disabledPlans?.[ plan.periodAgnosticSlug ]
 								}
-								isSelected={ plan.productIds.indexOf( selectedPlanProductId as never ) > -1 }
+								isSelected={
+									!! selectedPlanProductId &&
+									selectedPlanProductId ===
+										getPlanProduct( plan.periodAgnosticSlug, billingPeriod )?.productId
+								}
```
Same as https://github.com/Automattic/wp-calypso/pull/48622/files#r566841545

ciampo commented on packages/plans-grid/src/plans-table/index.tsx:
```diff
@@ -54,34 +56,39 @@ const PlansTable: React.FunctionComponent< Props > = ( {
-							onSelect={ onPlanSelect }
-							onPickDomainClick={ onPickDomainClick }
-							onToggleExpandAll={ () => setAllPlansExpanded( ( expand ) => ! expand ) }
-							disabledLabel={ disabledPlans?.[ plan.periodAgnosticSlug ] }
-						></PlanItem>
-					)
-			) }
+			{ supportedPlans
+				.filter( ( plan ) => !! plan )
+				.map( ( plan ) => (
+					<PlanItem
+						popularBadgeVariation={ popularBadgeVariation }
+						allPlansExpanded={ allPlansExpanded }
+						key={ plan.periodAgnosticSlug }
+						slug={ plan.periodAgnosticSlug }
+						domain={ currentDomain }
+						tagline={
+							( showTaglines && customTagLines?.[ plan.periodAgnosticSlug ] ) ?? plan.description
+						}
+						CTAVariation={ CTAVariation }
+						features={ plan.features ?? [] }
+						billingPeriod={ billingPeriod }
+						isPopular={ plan.isPopular }
+						isFree={ plan.isFree }
+						name={ plan?.title.toString() }
+						isSelected={
+							!! selectedPlanProductId &&
+							selectedPlanProductId ===
+								getPlanProduct( plan.periodAgnosticSlug, billingPeriod )?.productId
+						}
+						onSelect={ onPlanSelect }
+						onPickDomainClick={ onPickDomainClick }
+						onToggleExpandAll={ () => setAllPlansExpanded( ( expand ) => ! expand ) }
+						disabledLabel={ disabledPlans?.[ plan.periodAgnosticSlug ] }
+					></PlanItem>
+				) ) }
```
The big diff is because of a change of indentation. Only 2 things have changed:

1. Same as https://github.com/Automattic/wp-calypso/pull/48622/files#r566841545
2. added `.filter( ( plan ) => !! plan )` to avoid doing `plan && (<PlanItem...`

yansern reviewed (approved):
Smoke tested on focused launch (and also made sure gutenboarding, step-by-step didn't break). Good to go!
